### PR TITLE
Table calculations

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
Implemented table calculations

#### Ticket Link
Implements #461 

Calculations implemented -

1. count
1. countValue
1. countUniqueValue
1. sum
1. average
1. median
1. min
1. max
1. range

### CODE REVIEW

**Comment by hahmadia** [comment]
Yo Harshil,

Before doing code review I decided to test it out. Here are some minor things I found after a quick run through. 

When the column too small, the range gets stacked on top like this ( I think it should be come ellipsis)
image

For date, the unique should be 3 here right?
image

Count values for multi select should be 7 here right?
image


For Checkbox, shouldn't false values be accounted for too?
image

Unique values here should be 2 (I assume we account for false values too if I'm not wrong)
image

Created Time 0 values? Is this right?
image


Created time unique values also 0?
image

Created by value is 0?
image


Created by unique value is also 0?
image

**Review by hahmadia — CHANGES_REQUESTED** [review]
Found some minor bugs to address.

Note I didn't test every single property since some have overlap but you should check them all just to make sure I didn't miss something by accident. Thanks

**Comment by harshilsharma63** [comment]
Thanks @hahmadia . I'm working on these.

**Comment by harshilsharma63** [comment]
> When the column too small, the range gets stacked on top like this ( I think it should be come ellipsis)

Fixed.

> For date, the unique should be 3 here right?

Hmm, can you share an export of the board? Also can you verify it on the latest branch? I think it's fixed when I merged `main` in.

> Count values for multi select should be 7 here right?

Fixed

> For Checkbox, shouldn't false values be accounted for too?

Later on, we'll add checkbox specific values - `Checked` and `Unchecked`, those would make more sense. I think for now this is fine.

> Unique values here should be 2 (I assume we account for false values too if I'm not wrong)

Same as considering false values, later on, we'll add checkbox specific values - `Checked` and `Unchecked`, those would make more sense. I think for now this is fine.

> Created Time 0 values? Is this right?

Fixed

> Created time unique values also 0?

Fixed

> Created by value is 0?

Fixed

> Created by unique value is also 0?

Fixed

**Comment by hahmadia** [comment]
Hi Harshil,

Regarding todays meeting, I would be in favor of having this QA review after dev review.

**Comment by harshilsharma63** [comment]
@hahmadia you can review this now. I've synced this with `main`.

Agree on the QA point.

**Comment by hahmadia** [comment]
Hi Harshil

Menu gets hidden on numbers. Also recommend adding "..." to range so we can know we should expand. I thought it was a bug but realized column was too small.

image

---

Have date like this

image

Delete first value

image

Change first value to '08/13/2021'
image

Unique value stays at 2

**Changing it in the card using the calendar works fine and doesn't cause this problem**

---

Shouldn't unique only be 4 for created time here? 

image

---

Last updated time unique should be 2 here right?
image



Otherwise nice job! Everything else seems to be working. I won't wait for you to fix these to not hold you back. I'll go forward with doing code review now.

File: webapp/src/components/calculations/calculations.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {Card} from '../../blocks/card'
+import {IPropertyTemplate} from '../../blocks/board'
+import {Utils} from '../../utils'
+import {Constants} from '../../constants'
+
+const roundedDecimalPlaces = 2
```
Comment (hahmadia) [inline webapp/src/components/calculations/calculations.ts:9]: Probably wanna make this all caps

image

**Review by hahmadia — APPROVED** [review]
Looks good to me with the exception of incorrect calculations

**Comment by ogi-m** [comment]
@hahmadia  Regarding the unique count on the dates:
If there's 3 dates in the column - 2x`August 13` and 1x`August 31` and `Unique` count shows 2 - isn't that correct?

**Comment by hahmadia** [comment]
> @hahmadia  Regarding the unique count on the dates:
> 
> If there's 3 dates in the column - 2x`August 13` and 1x`August 31` and `Unique` count shows 2 - isn't that correct?

@ogi-m 

Yes sorry! That is correct.

**Review by ogi-m — APPROVED** [review]

**Comment by chenilim** [comment]
@metanerd (I think you still own mattermod?), do you know if slash cherry-pick should work on this repo? If so, is there something that needs to be configured? Thanks.

**Pull request merged**